service/src/main/java/bio/terra/workspace/service/iam/SamService.java

```diff
@@ -1219,6 +1219,42 @@ public Optional<ServiceAccountName> constructUserPetSaEmail(
     }
   }
 
+  /** List Sam resource IDs for the calling user for the given resource type. */
+  public List<String> listResourceIds(AuthenticatedUserRequest userRequest, String resourceType)
+      throws InterruptedException {
+    ResourcesApi resourcesApi = samResourcesApi((userRequest.getRequiredToken()));
+    try {
+      List<UserResourcesResponse> resources =
+          SamRetry.retry(() -> resourcesApi.listResourcesAndPoliciesV2(resourceType));
+      return resources.stream()
+          .map(UserResourcesResponse::getResourceId)
+          .collect(Collectors.toList());
+    } catch (ApiException apiException) {
+      throw SamExceptionFactory.create("Error listing resources from Sam", apiException);
+    }
+  }
+
+  /**
+   * Fetch an Action Managed Identity (if present) for the calling user for the given Sam resource
+   * type, id, and action.
+   */
+  public Optional<String> getActionManagedIdentity(
+      AuthenticatedUserRequest userRequest, String resourceType, String resourceId, String action)
+      throws InterruptedException {
+    AzureApi azureApi = samAzureApi(userRequest.getRequiredToken());
```

Comment:
There is an open question here. `userRequest.getRequiredToken()` actually maybe represent the Leo SA (since VMs are application-owned resources). It's unclear if the Sam AAMI resource should be assigned to the Leo SA, the WSM SA, or we should invent a more administrative API in Sam to get an AAMI on behalf of a user.